File: backend/app/schemas/wrong_book.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator


class WrongBookItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    subject_code: str
    knowledge_node_id: uuid.UUID | None
    source_type: str
    source_id: uuid.UUID | None
    question_snapshot_json: dict
    answer_snapshot_json: dict
    correct_snapshot_json: dict
    status: str
    wrong_count: int
    consecutive_correct_count: int
    mastered_at: datetime | None
    last_practice_at: datetime | None
    created_at: datetime
    has_explanation: bool = False

    @model_validator(mode="wrap")
    @classmethod
    def _has_explanation_from_orm(cls, data, handler):
        if hasattr(data, "explanation_text"):
            out = handler(data)
            return out.model_copy(
                update={"has_explanation": bool(data.explanation_text)}
            )
        return handler(data)


class WrongBookPracticeIn(BaseModel):
    content: str = Field(min_length=1, max_length=8000)


class WrongBookPracticeOut(BaseModel):
    is_correct: bool
    status: str
    consecutive_correct_count: int
    mastered: bool


class WrongBookExplainIn(BaseModel):
    regenerate: bool = False


class WrongBookExplainOut(BaseModel):
    explanation_text: str
    from_cache: bool
    explanation_created_at: datetime | None

File: frontend/src/api/wrongBook.ts
import { api } from "./client";

export type WrongBookItemOut = {
  id: string;
  subject_code: string;
  knowledge_node_id: string | null;
  source_type: string;
  source_id: string | null;
  question_snapshot_json: Record<string, unknown>;
  answer_snapshot_json: Record<string, unknown>;
  correct_snapshot_json: Record<string, unknown>;
  status: string;
  wrong_count: number;
  consecutive_correct_count: number;
  mastered_at: string | null;
  last_practice_at: string | null;
  created_at: string;
};

export type WrongBookPracticeOut = {
  is_correct: boolean;
  status: string;
  consecutive_correct_count: number;
  mastered: boolean;
};

export function listWrongBook(params?: {
  subject_code?: string;
  source_type?: string;
  knowledge_node_id?: string;
  status?: string;
  limit?: number;
  offset?: number;
}) {
  const qsParams = new URLSearchParams();
  if (params?.subject_code) qsParams.set("subject_code", params.subject_code);
  if (params?.source_type) qsParams.set("source_type", params.source_type);
  if (params?.knowledge_node_id) qsParams.set("knowledge_node_id", params.knowledge_node_id);
  if (params?.status) qsParams.set("status", params.status);
  if (typeof params?.limit === "number") qsParams.set("limit", String(params.limit));
  if (typeof params?.offset === "number") qsParams.set("offset", String(params.offset));
  const qs = qsParams.toString() ? `?${qsParams.toString()}` : "";
  return api<WrongBookItemOut[]>(`/student/wrong-book${qs}`);
}

export function practiceWrongItem(itemId: string, content: string) {
  return api<WrongBookPracticeOut>(`/student/wrong-book/${itemId}/practice`, {
    method: "POST",
    body: JSON.stringify({ content }),
  });
}

export function archiveWrongItem(itemId: string) {
  return api<WrongBookItemOut>(`/student/wrong-book/${itemId}/archive`, {
    method: "POST",
  });
}
